File: py/packages/halo/src/halo_format/hash.py
"""canonical bytes -> handle.

handle = "h:" + sha256(canonical(node)).hexdigest(), full 64-hex. Algorithm pluggable behind a
small registry keyed by alg (default sha256); the envelope declares it per-tree. Uses stdlib
hashlib. Leaf dependency, byte-exact across ports.
"""

import hashlib

from .canonical import canonical_bytes

# Algorithm registry keyed by alg. A future move (e.g. blake3) registers a function here and the
# envelope's alg selects it; call sites never hard-code the algorithm.
_REGISTRY = {
    "sha256": lambda b: hashlib.sha256(b).hexdigest(),
}


def hash_bytes(data: bytes, alg: str = "sha256") -> str:
    """Hash canonical bytes into a handle: "h:" + lowercase hex digest under alg."""
    fn = _REGISTRY.get(alg)
    if fn is None:
        raise ValueError(f"unknown hash alg: {alg}")
    return "h:" + fn(data)


def handle_of(value, alg: str = "sha256") -> str:
    """Convenience: handle of a JSON value = hash_bytes(canonical_bytes(value))."""
    return hash_bytes(canonical_bytes(value), alg)

File: ts/packages/halo/src/hash.ts
// hash: canonical bytes -> handle.
//
// handle = "h:" + lowercaseHex(sha256(canonical(node))), full 64-hex digest. Algorithm is
// pluggable behind a tiny registry keyed by `alg` (default sha256); the envelope declares
// it per-tree. Leaf dependency, byte-exact across ports.

import { createHash } from "node:crypto";
import type { Alg, Handle, JsonValue } from "./types.js";
import { canonicalBytes } from "./canonical.js";

// Algorithm registry keyed by `alg`. A future move (e.g. blake3) registers a function here and
// the envelope's `alg` selects it; call sites never hard-code the algorithm.
const REGISTRY: Record<Alg, (bytes: Uint8Array) => string> = {
  sha256: (bytes) => createHash("sha256").update(bytes).digest("hex"),
};

/** Hash canonical bytes into a handle: "h:" + lowercase hex digest under `alg`. */
export function hashBytes(bytes: Uint8Array, alg: Alg = "sha256"): Handle {
  const fn = REGISTRY[alg];
  if (!fn) throw new Error(`unknown hash alg: ${alg}`);
  return "h:" + fn(bytes);
}

/** Convenience: handle of a JSON value = hashBytes(canonicalBytes(value)). */
export function handleOf(value: JsonValue, alg: Alg = "sha256"): Handle {
  return hashBytes(canonicalBytes(value), alg);
}
